File: eve_mcp/tools/character.py
"""Character sheet: skills, training queue, clones, standings."""
from __future__ import annotations

import asyncio
from datetime import datetime, timezone
from typing import Annotated, Any

from mcp.server.mcpserver import MCPServer
from pydantic import Field

from ..context import AppContext
from ..schema import CharacterArg, DetailArg, limit_arg
from ._common import handled, page, project

_SkillsLimit = limit_arg("skills", 400)
_StandingsEntriesLimit = limit_arg("standings entries", 300)

_ROMAN = {0: "0", 1: "I", 2: "II", 3: "III", 4: "IV", 5: "V"}


def register(mcp: MCPServer, ctx: AppContext) -> None:
    @mcp.tool()
    @handled
    async def eve_character_skills(
        character: CharacterArg = "",
        search: Annotated[
            str,
            Field(
                description=(
                    "Case-insensitive substring of the skill name, e.g. 'Gunnery' or "
                    "'Caldari'. Strongly recommended — a full skill list is hundreds "
                    "of rows and answers almost nothing on its own."
                )
            ),
        ] = "",
        trained_only: Annotated[
            bool,
            Field(
                description=(
                    "Hide skills that are injected but sitting at level 0. Turn off "
                    "only when checking whether a skill book has been injected at all."
                )
            ),
        ] = True,
        limit: _SkillsLimit = 20,
        response_format: DetailArg = "concise",
    ) -> dict[str, Any]:
        """Trained skills with levels and skill points.

        Prefer `search` over dumping everything: to answer "can I fly a Drake"
        you want the handful of relevant skills, not all 118.

        One subtlety worth surfacing to the user: `active_level` can be lower
        than `level`. That means the account is on an Alpha (free) clone, which
        caps many skills below what has actually been trained. If you see the
        two diverge, say so — the character is not getting the benefit of
        training they already paid for.

        Returns: total_sp, unallocated_sp, skills_known, at_level_5, skills[].
        'detailed' adds skillpoints and active_level per skill.
        """
        token = ctx.resolve_character(character)
        cid = token.character_id
        ctx.require_scope(token, "esi-skills.read_skills.v1", "skills")
        concise = response_format == "concise"

        result = await ctx.esi.get(f"/characters/{cid}/skills", character_id=cid)
        payload = result.data or {}
        skills = [s for s in payload.get("skills", []) if isinstance(s, dict)]
        names = await ctx.resolver.names({s["skill_id"] for s in skills})

        needle = search.strip().lower()
        rows = []
        capped = 0
        for skill in skills:
            name = names.get(skill["skill_id"], f"#{skill['skill_id']}")
            level = skill.get("trained_skill_level", 0)
            active = skill.get("active_skill_level", level)
            if active < level:
                capped += 1
            if trained_only and not level:
                continue
            if needle and needle not in name.lower():
                continue
            rows.append(
                {
                    "skill": name,
                    "level": _ROMAN.get(level, str(level)),
                    "skillpoints": skill.get("skillpoints_in_skill"),
                    "active_level": _ROMAN.get(active, str(active)),
                }
            )
        rows.sort(key=lambda r: (r["skill"]))
        visible, meta = page(rows, limit, "Narrow with `search`, or raise `limit`.")

        out = {
            "character": token.character_name,
            "total_sp": payload.get("total_sp"),
            "unallocated_sp": payload.get("unallocated_sp"),
            "skills_known": len(skills),
            "at_level_5": sum(1 for s in skills if s.get("trained_skill_level") == 5),
            "matching": len(rows),
            "data_age": result.stale_note,
            **meta,
            "skills": project(visible, ("skill", "level"), concise),
        }
        if capped:
            out["alpha_clone_warning"] = (
                f"{capped} skills have active_level below trained level — this account "
                "looks like it is on an Alpha clone, so trained levels are capped."
            )
        return out

    @mcp.tool()
    @handled
    async def eve_character_skill_queue(character: CharacterArg = "") -> dict[str, Any]:
        """The training queue: what is training now, what follows, and when it runs dry.

        An empty queue means the character is accruing nothing — always worth
        telling the user, because it is silent in game and easy to miss.

        Cheap to call (a dozen rows at most), so prefer it over eve_character_skills
        when the question is about training rather than what is already known.

        Returns: queued_skills, training_now, queue_empty_in, queue_ends, queue[].
        """
        token = ctx.resolve_character(character)
        cid = token.character_id
        ctx.require_scope(token, "esi-skills.read_skillqueue.v1", "the skill queue")

        result = await ctx.esi.get(f"/characters/{cid}/skillqueue", character_id=cid)
        entries = [e for e in (result.data or []) if isinstance(e, dict)]
        if not entries:
            return {
                "character": token.character_name,
                "queue": [],
                "warning": (
                    "The skill queue is empty. This character is accruing no skill "
                    "points at all until something is queued."
                ),
            }
        names = await ctx.resolver.names({e["skill_id"] for e in entries})
        now = datetime.now(timezone.utc)

        rows = []
        for entry in sorted(entries, key=lambda e: e.get("queue_position", 0)):
            finish = _parse(entry.get("finish_date"))
            rows.append(
                {
                    "position": entry.get("queue_position"),
                    "skill": names.get(entry["skill_id"], f"#{entry['skill_id']}"),
                    "to_level": _ROMAN.get(entry.get("finished_level", 0)),
                    "finishes_in": _human_delta(finish - now) if finish else "paused",
                    "finish_date": entry.get("finish_date"),
                }
            )
        last = _parse(rows[-1]["finish_date"]) if rows[-1]["finish_date"] else None
        return {
            "character": token.character_name,
            "queued_skills": len(rows),
            "training_now": f"{rows[0]['skill']} {rows[0]['to_level'] or ''}".strip(),
            "queue_empty_in": _human_delta(last - now) if last else "unknown",
            "queue_ends": rows[-1]["finish_date"],
            "data_age": result.stale_note,
            "queue": rows,
        }

    @mcp.tool()
    @handled
    async def eve_character_clones(character: CharacterArg = "") -> dict[str, Any]:
        """Jump clones with their locations and implants, plus the active clone's implants.

        Useful for "where can I jump to" and "what implants would I lose if I
        died right now" — implants in the active clone are destroyed on podding,
        the ones sitting in jump clones are not.

        Returns: home_station, last_clone_jump, active_implants[], jump_clones[].
        """
        token = ctx.resolve_character(character)
        cid = token.character_id
        ctx.require_scope(token, "esi-clones.read_clones.v1", "clones")
        # The implants call is a second endpoint with a second scope; without
        # this check it fails the whole tool with an unexplained 403.
        ctx.require_scope(token, "esi-clones.read_implants.v1", "the active clone's implants")

        clones_result, implants_result = await asyncio.gather(
            ctx.esi.get(f"/characters/{cid}/clones", character_id=cid),
            ctx.esi.get(f"/characters/{cid}/implants", character_id=cid),
        )
        clones = clones_result.data or {}
        active_implants = implants_result.data or []

        jump_clones = clones.get("jump_clones", [])
        implant_ids = set(active_implants)
        location_ids = set()
        for clone in jump_clones:
            implant_ids.update(clone.get("implants", []))
            location_ids.add(clone.get("location_id"))
        home = clones.get("home_location", {})
        if home.get("location_id"):
            location_ids.add(home["location_id"])

        names = await ctx.resolver.names(implant_ids | location_ids, character_id=cid)
        return {
            "character": token.character_name,
            "home_station": names.get(home.get("location_id")),
            "last_clone_jump": clones.get("last_clone_jump_date"),
            "active_implants": [names.get(i, f"#{i}") for i in active_implants],
            "jump_clones": [
                {
                    "name": clone.get("name") or f"Clone {clone.get('jump_clone_id')}",
                    "location": names.get(clone.get("location_id"), "unknown"),
                    "implants": [names.get(i, f"#{i}") for i in clone.get("implants", [])],
                }
                for clone in jump_clones
            ],
            "data_age": clones_result.stale_note,
        }

    @mcp.tool()
    @handled
    async def eve_character_standings(
        character: CharacterArg = "",
        limit: _StandingsEntriesLimit = 20,
    ) -> dict[str, Any]:
        """NPC faction and corporation standings, plus loyalty point balances.

        Standings run -10 to +10 and gate agent access, broker fees and whether
        a faction's navy shoots you. Loyalty points are the currency earned from
        that faction's missions, spendable in their LP store.

        Returns: loyalty_points[], standings[] sorted best-first.
        """
        token = ctx.resolve_character(character)
        cid = token.character_id
        ctx.require_scope(token, "esi-characters.read_standings.v1", "standings")
        # Loyalty points are a separate endpoint with a separate scope. Missing it
        # is not fatal — standings still answer — so note it instead of raising.
        lp_scope = "esi-characters.read_loyalty.v1"
        lp_granted = lp_scope in token.scopes

        calls = [ctx.esi.get(f"/characters/{cid}/standings", character_id=cid)]
        if lp_granted:
            # Skipping the call when the scope is absent avoids spending a
            # guaranteed 403 against the shared ESI error budget every time.
            calls.append(ctx.esi.get(f"/characters/{cid}/loyalty/points", character_id=cid))
        results = await asyncio.gather(*calls, return_exceptions=True)
        standings_result = results[0]
        lp_result = results[1] if lp_granted else None
        standings = getattr(standings_result, "data", []) or []
        loyalty = getattr(lp_result, "data", []) or []

        ids = {s["from_id"] for s in standings} | {l["corporation_id"] for l in loyalty}
        names = await ctx.resolver.names(ids)

        rows = sorted(
            (
                {
                    "entity": names.get(s["from_id"], f"#{s['from_id']}"),
                    "type": s.get("from_type"),
                    "standing": round(s.get("standing", 0.0), 2),
                }
                for s in standings
            ),
            key=lambda r: -r["standing"],
        )
        visible, meta = page(rows, limit)
        out = {
            "character": token.character_name,
            "loyalty_points": [
                {"corporation": names.get(l["corporation_id"]), "lp": l.get("loyalty_points")}
                for l in sorted(loyalty, key=lambda l: -l.get("loyalty_points", 0))
            ],
            **meta,
            "standings": visible,
        }
        if isinstance(standings_result, Exception):
            # An empty list here would be indistinguishable from "no standings".
            out["standings_note"] = (
                f"Standings could not be read: {standings_result}. The list above is "
                "empty because the call failed, not because there are none."
            )
        if isinstance(lp_result, Exception):
            out["loyalty_points_note"] = (
                f"Loyalty points could not be read: {lp_result}."
            )
        elif not lp_granted:
            out["loyalty_points_note"] = (
                f"{token.character_name} was not authorized with '{lp_scope}', so "
                "loyalty point balances are not available. Re-run the login for this "
                "character to include them."
            )
        return out


def _parse(value: str | None) -> datetime | None:
    if not value:
        return None
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None


def _human_delta(delta) -> str:
    seconds = int(delta.total_seconds())
    if seconds < 0:
        return "done"
    days, rem = divmod(seconds, 86400)
    hours, rem = divmod(rem, 3600)
    minutes = rem // 60
    if days:
        return f"{days}d {hours}h"
    if hours:
        return f"{hours}h {minutes}m"
    return f"{minutes}m"

File: internal/usecase/eve/character.go
package eve

import (
	"context"
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/truewebber/eve-online-mcp/internal/domain/j"
	"github.com/truewebber/eve-online-mcp/internal/usecase/session"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

type characterSkillsIn struct {
	Character      string `json:"character,omitempty"       jsonschema:"Character name (e.g. 'Jane Doe') or numeric character id. Leave empty to use the only authorized character; required when several are authorized — call eve_auth_status to list them."`
	Search         string `json:"search,omitempty"          jsonschema:"Case-insensitive substring of the skill name, e.g. 'Gunnery' or 'Caldari'. Strongly recommended — a full skill list is hundreds of rows."`
	TrainedOnly    *bool  `json:"trained_only,omitempty"    jsonschema:"Hide skills that are injected but sitting at level 0. Default true."`
	Limit          int    `json:"limit,omitempty"           jsonschema:"Maximum rows to return. Keep it small — every row costs context. Results say truncated when more exist."`
	ResponseFormat string `json:"response_format,omitempty" jsonschema:"'concise' (default) returns only the high-signal fields and costs far fewer tokens. Use 'detailed' when you need secondary fields and raw ids."`
}

type characterSkillQueueIn struct {
	Character string `json:"character,omitempty" jsonschema:"Character name (e.g. 'Jane Doe') or numeric character id. Leave empty to use the only authorized character; required when several are authorized — call eve_auth_status to list them."`
}

type characterClonesIn struct {
	Character string `json:"character,omitempty" jsonschema:"Character name (e.g. 'Jane Doe') or numeric character id. Leave empty to use the only authorized character; required when several are authorized — call eve_auth_status to list them."`
}

type characterStandingsIn struct {
	Character string `json:"character,omitempty" jsonschema:"Character name (e.g. 'Jane Doe') or numeric character id. Leave empty to use the only authorized character; required when several are authorized — call eve_auth_status to list them."`
	Limit     int    `json:"limit,omitempty"     jsonschema:"Maximum rows to return. Keep it small — every row costs context. Results say truncated when more exist."`
}

func registerCharacter(s *mcp.Server) {
	addTool(s, &mcp.Tool{
		Name:        "eve_character_skills",
		Description: "Trained skills with levels and skill points.\n\nPrefer `search` over dumping everything: to answer \"can I fly a Drake\" you want the handful of relevant skills, not all 118.\n\nOne subtlety worth surfacing to the user: `active_level` can be lower than `level`. That means the account is on an Alpha (free) clone.\n\nReturns: total_sp, unallocated_sp, skills_known, at_level_5, skills[].",
	}, sessionTool(eveCharacterSkills))
	addTool(s, &mcp.Tool{
		Name:        "eve_character_skill_queue",
		Description: "The training queue: what is training now, what follows, and when it runs dry.\n\nAn empty queue means the character is accruing nothing — always worth telling the user.\n\nReturns: queued_skills, training_now, queue_empty_in, queue_ends, queue[].",
	}, sessionTool(eveCharacterSkillQueue))
	addTool(s, &mcp.Tool{
		Name:        "eve_character_clones",
		Description: "Jump clones with their locations and implants, plus the active clone's implants.\n\nUseful for \"where can I jump to\" and \"what implants would I lose if I died right now\".\n\nReturns: home_station, last_clone_jump, active_implants[], jump_clones[].",
	}, sessionTool(eveCharacterClones))
	addTool(s, &mcp.Tool{
		Name:        "eve_character_standings",
		Description: "NPC faction and corporation standings, plus loyalty point balances.\n\nStandings run -10 to +10 and gate agent access, broker fees and whether a faction's navy shoots you.\n\nReturns: loyalty_points[], standings[] sorted best-first.",
	}, sessionTool(eveCharacterStandings))
}

func eveCharacterSkills(ctx context.Context, a *session.Session, in characterSkillsIn) (any, error) {
	token, err := a.ResolveCharacter(ctx, in.Character)
	if err != nil {
		return nil, err
	}
	if err := a.RequireScope(token, "esi-skills.read_skills.v1", "skills"); err != nil {
		return nil, err
	}
	cid := token.CharacterID
	result, err := a.ESI.Get(ctx, fmt.Sprintf("/characters/%d/skills", cid), &cid, nil, nil)
	if err != nil {
		return nil, err
	}
	payload := j.Map(result.Data)
	skills := j.Maps(payload["skills"])
	var ids []int
	for _, s := range skills {
		ids = append(ids, j.Int(s["skill_id"]))
	}
	names, err := a.Resolver.Names(ctx, ids, nil)
	if err != nil {
		return nil, err
	}
	view := filterCharacterSkills(skills, names, in.Search, boolDef(in.TrainedOnly, true))
	sort.Slice(view.rows, func(i, k int) bool { return j.Str(view.rows[i]["skill"]) < j.Str(view.rows[k]["skill"]) })
	visible, meta := page(view.rows, limitOr(in.Limit, 20), "Narrow with `search`, or raise `limit`.")
	at5 := 0
	for _, s := range skills {
		if j.Int(s["trained_skill_level"]) == 5 {
			at5++
		}
	}
	out := merge(map[string]any{
		"character": token.CharacterName, "total_sp": payload["total_sp"],
		"unallocated_sp": payload["unallocated_sp"], "skills_known": len(skills),
		"at_level_5": at5, "matching": len(view.rows), "data_age": result.StaleNote(),
		"skills": project(visible, []string{"skill", "level"}, concise(in.ResponseFormat)),
	}, meta)
	if view.capped > 0 {
		out["alpha_clone_warning"] = fmt.Sprintf("%d skills have active_level below trained level — this account looks like it is on an Alpha clone, so trained levels are capped.", view.capped)
	}

	return out, nil
}

type characterSkillsView struct {
	rows   []map[string]any
	capped int
}

func filterCharacterSkills(skills []map[string]any, names map[int]string, search string, trainedOnly bool) characterSkillsView {
	needle := strings.ToLower(strings.TrimSpace(search))
	var rows []map[string]any
	capped := 0
	for _, skill := range skills {
		name := names[j.Int(skill["skill_id"])]
		if name == "" {
			name = fmt.Sprintf("#%d", j.Int(skill["skill_id"]))
		}
		level := j.Int(skill["trained_skill_level"])
		active := j.Int(skill["active_skill_level"])
		if active == 0 && skill["active_skill_level"] == nil {
			active = level
		}
		if active < level {
			capped++
		}
		if !keepCharacterSkill(name, level, needle, trainedOnly) {
			continue
		}
		rows = append(rows, map[string]any{
			"skill": name, "level": roman(level),
			"skillpoints": skill["skillpoints_in_skill"], "active_level": roman(active),
		})
	}

	return characterSkillsView{rows: rows, capped: capped}
}

func keepCharacterSkill(name string, level int, needle string, trainedOnly bool) bool {
	if trainedOnly && level == 0 {
		return false
	}
	if needle != "" && !strings.Contains(strings.ToLower(name), needle) {
		return false
	}

	return true
}

func eveCharacterSkillQueue(ctx context.Context, a *session.Session, in characterSkillQueueIn) (any, error) {
	token, err := a.ResolveCharacter(ctx, in.Character)
	if err != nil {
		return nil, err
	}
	if err := a.RequireScope(token, "esi-skills.read_skillqueue.v1", "the skill queue"); err != nil {
		return nil, err
	}
	cid := token.CharacterID
	result, err := a.ESI.Get(ctx, fmt.Sprintf("/characters/%d/skillqueue", cid), &cid, nil, nil)
	if err != nil {
		return nil, err
	}
	entries := j.Maps(result.Data)
	if len(entries) == 0 {
		return map[string]any{
			"character": token.CharacterName, "queue": []any{},
			"warning": "The skill queue is empty. This character is accruing no skill points at all until something is queued.",
		}, nil
	}
	var ids []int
	for _, e := range entries {
		ids = append(ids, j.Int(e["skill_id"]))
	}
	names, err := a.Resolver.Names(ctx, ids, nil)
	if err != nil {
		return nil, err
	}
	now := time.Now().UTC()
	sort.Slice(entries, func(i, k int) bool { return j.Int(entries[i]["queue_position"]) < j.Int(entries[k]["queue_position"]) })
	rows := formatSkillQueue(entries, names, now)
	emptyIn := "unknown"
	if last := parseTime(j.Str(rows[len(rows)-1]["finish_date"])); last != nil {
		emptyIn = humanDelta(last.Sub(now))
	}

	return map[string]any{
		"character": token.CharacterName, "queued_skills": len(rows),
		"training_now":   strings.TrimSpace(j.Str(rows[0]["skill"]) + " " + j.Str(rows[0]["to_level"])),
		"queue_empty_in": emptyIn, "queue_ends": rows[len(rows)-1]["finish_date"],
		"data_age": result.StaleNote(), "queue": rows,
	}, nil
}

func formatSkillQueue(entries []map[string]any, names map[int]string, now time.Time) []map[string]any {
	rows := make([]map[string]any, 0, len(entries))
	for _, e := range entries {
		finish := parseTime(j.Str(e["finish_date"]))
		finishes := "paused"
		if finish != nil {
			finishes = humanDelta(finish.Sub(now))
		}
		name := names[j.Int(e["skill_id"])]
		if name == "" {
			name = fmt.Sprintf("#%d", j.Int(e["skill_id"]))
		}
		rows = append(rows, map[string]any{
			"position": e["queue_position"], "skill": name,
			"to_level": roman(j.Int(e["finished_level"])), "finishes_in": finishes,
			"finish_date": e["finish_date"],
		})
	}

	return rows
}

func eveCharacterClones(ctx context.Context, a *session.Session, in characterClonesIn) (any, error) {
	token, err := a.ResolveCharacter(ctx, in.Character)
	if err != nil {
		return nil, err
	}
	if err := a.RequireScope(token, "esi-clones.read_clones.v1", "clones"); err != nil {
		return nil, err
	}
	if err := a.RequireScope(token, "esi-clones.read_implants.v1", "the active clone's implants"); err != nil {
		return nil, err
	}
	cid := token.CharacterID
	clonesRes, err := a.ESI.Get(ctx, fmt.Sprintf("/characters/%d/clones", cid), &cid, nil, nil)
	if err != nil {
		return nil, err
	}
	implantsRes, err := a.ESI.Get(ctx, fmt.Sprintf("/characters/%d/implants", cid), &cid, nil, nil)
	if err != nil {
		return nil, err
	}
	clones := j.Map(clonesRes.Data)
	implants := j.Slice(implantsRes.Data)
	jump := j.Maps(clones["jump_clones"])
	home := j.Map(clones["home_location"])
	names, err := a.Resolver.Names(ctx, collectCloneIDs(implants, jump, home), &cid)
	if err != nil {
		return nil, err
	}

	return map[string]any{
		"character":       token.CharacterName,
		"home_station":    names[j.Int(home["location_id"])],
		"last_clone_jump": clones["last_clone_jump_date"],
		"active_implants": formatActiveImplants(implants, names), "jump_clones": formatJumpClones(jump, names),
		"data_age": clonesRes.StaleNote(),
	}, nil
}

func collectCloneIDs(implants []any, jump []map[string]any, home map[string]any) []int {
	idSet := map[int]struct{}{}
	for _, v := range implants {
		idSet[j.Int(v)] = struct{}{}
	}
	for _, clone := range jump {
		for _, v := range j.Slice(clone["implants"]) {
			idSet[j.Int(v)] = struct{}{}
		}
		idSet[j.Int(clone["location_id"])] = struct{}{}
	}
	if j.Int(home["location_id"]) != 0 {
		idSet[j.Int(home["location_id"])] = struct{}{}
	}

	return setToList(idSet)
}

func formatActiveImplants(implants []any, names map[int]string) []string {
	active := make([]string, 0, len(implants))
	for _, v := range implants {
		active = append(active, names[j.Int(v)])
	}

	return active
}

func formatJumpClones(jump []map[string]any, names map[int]string) []map[string]any {
	jumps := make([]map[string]any, 0, len(jump))
	for _, clone := range jump {
		listed := j.Slice(clone["implants"])
		imps := make([]string, 0, len(listed))
		for _, v := range listed {
			imps = append(imps, names[j.Int(v)])
		}
		name := j.Str(clone["name"])
		if name == "" {
			name = fmt.Sprintf("Clone %v", clone["jump_clone_id"])
		}
		loc := names[j.Int(clone["location_id"])]
		if loc == "" {
			loc = "unknown"
		}
		jumps = append(jumps, map[string]any{"name": name, "location": loc, "implants": imps})
	}

	return jumps
}

func eveCharacterStandings(ctx context.Context, a *session.Session, in characterStandingsIn) (any, error) {
	token, err := a.ResolveCharacter(ctx, in.Character)
	if err != nil {
		return nil, err
	}
	if err := a.RequireScope(token, "esi-characters.read_standings.v1", "standings"); err != nil {
		return nil, err
	}
	cid := token.CharacterID
	standingsRes, standingsErr := a.ESI.Get(ctx, fmt.Sprintf("/characters/%d/standings", cid), &cid, nil, nil)
	lpScope := "esi-characters.read_loyalty.v1"
	lpGranted := a.HasScope(token, lpScope)
	lpData, lpErr := fetchCharacterLP(ctx, a, cid, lpGranted)
	var standings []map[string]any
	if standingsErr == nil {
		standings = j.Maps(standingsRes.Data)
	}
	names, err := a.Resolver.Names(ctx, standingsNameIDs(standings, lpData), nil)
	if err != nil {
		return nil, err
	}
	rows := formatCharacterStandings(standings, names)
	visible, meta := page(rows, limitOr(in.Limit, 20), "")
	out := merge(map[string]any{
		"character": token.CharacterName, "loyalty_points": formatLoyaltyPoints(lpData, names), "standings": visible,
	}, meta)
	applyStandingsNotes(out, standingsErr, lpErr, lpGranted, token.CharacterName, lpScope)

	return out, nil
}

func fetchCharacterLP(ctx context.Context, a *session.Session, cid int, granted bool) ([]map[string]any, error) {
	if !granted {
		return nil, nil
	}
	lpRes, err := a.ESI.Get(ctx, fmt.Sprintf("/characters/%d/loyalty/points", cid), &cid, nil, nil)
	if err != nil {
		return nil, err
	}

	return j.Maps(lpRes.Data), nil
}

func standingsNameIDs(standings, lpData []map[string]any) []int {
	idSet := map[int]struct{}{}
	for _, s := range standings {
		idSet[j.Int(s["from_id"])] = struct{}{}
	}
	for _, l := range lpData {
		idSet[j.Int(l["corporation_id"])] = struct{}{}
	}

	return setToList(idSet)
}

func formatCharacterStandings(standings []map[string]any, names map[int]string) []map[string]any {
	rows := make([]map[string]any, 0, len(standings))
	for _, s := range standings {
		rows = append(rows, map[string]any{
			"entity": names[j.Int(s["from_id"])], "type": s["from_type"],
			"standing": mathRound(j.Float(s["standing"]), 2),
		})
	}
	sort.Slice(rows, func(i, k int) bool { return j.Float(rows[i]["standing"]) > j.Float(rows[k]["standing"]) })

	return rows
}

func formatLoyaltyPoints(lpData []map[string]any, names map[int]string) []map[string]any {
	sort.Slice(lpData, func(i, k int) bool {
		return j.Float(lpData[i]["loyalty_points"]) > j.Float(lpData[k]["loyalty_points"])
	})
	lpRows := make([]map[string]any, 0, len(lpData))
	for _, l := range lpData {
		lpRows = append(lpRows, map[string]any{"corporation": names[j.Int(l["corporation_id"])], "lp": l["loyalty_points"]})
	}

	return lpRows
}

func applyStandingsNotes(out map[string]any, standingsErr, lpErr error, lpGranted bool, character, lpScope string) {
	if standingsErr != nil {
		out["standings_note"] = fmt.Sprintf("Standings could not be read: %v. The list above is empty because the call failed, not because there are none.", standingsErr)
	}
	if lpErr != nil {
		out["loyalty_points_note"] = fmt.Sprintf("Loyalty points could not be read: %v.", lpErr)
	} else if !lpGranted {
		out["loyalty_points_note"] = fmt.Sprintf("%s was not authorized with '%s', so loyalty point balances are not available. Re-run the login for this character to include them.", character, lpScope)
	}
}
